### PR TITLE
Feature/452 transport params

### PR DESCRIPTION
**Pull request by marto83**
Spec align. Add custom Transport params dictionary to ClientOptions

### CODE REVIEW

File: src/IO.Ably.Shared/ClientOptions.cs [hunk 1]
```diff
@@ -296,6 +296,11 @@ public bool UseBinaryProtocol
         /// </summary>
         public bool IdempotentRestPublishing { get; set; } = Defaults.ProtocolVersionNumber >= 1.2;
 
+        /// <summary>
+        /// Additional parameters to be sent in the querystring when initiating a realtime connection.
+        /// </summary>
+        public Dictionary<string, string> TransportParams { get; set; } = new Dictionary<string, string>();
```
Comment (QuintinWillison) [inline src/IO.Ably.Shared/ClientOptions.cs:302]: [RTC1f](https://docs.ably.io/client-lib-development-guide/features/#RTC1f) states that these values should simply be `Stringifiable` (spec. wording, not mine) so perhaps the values here should be `Object` and your implementation then uses [`ToString()`](https://docs.microsoft.com/en-us/dotnet/api/system.object.tostring?view=netcore-3.1)?

**Review by QuintinWillison — CHANGES_REQUESTED** [review]

File: src/IO.Ably.Shared/ClientOptions.cs (hunk 1 above)
Comment (marto83) [reply src/IO.Ably.Shared/ClientOptions.cs:302]: I did see that but didn't think it added much value as these will be used for query parameters. I'm happy to change it and call `ToString()` if you think it's required.

Comment (sacOO7) [reply src/IO.Ably.Shared/ClientOptions.cs:302]: I agree with @QuintinWillison [RTC1f](https://docs.ably.io/client-lib-development-guide/features/#RTC1f), it states If values are unidiomatic to the language, the implementer may consider this as equivalent to String. Since dotnet is a typed language, it will be great if we can leverage that feature and be specific with values stored in the map. I am not sure if the current implementation works without being specific with values, I am considering usability wrt passed values in ClientOptions.

Comment (sacOO7) [reply src/IO.Ably.Shared/ClientOptions.cs:302]: Maybe use something like ```where T : struct, string``` for value

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSpecs/ConnectionParameterSpecs.cs [hunk 2]
```diff
@@ -137,6 +138,48 @@ public async Task ShouldSetTransportLibVersionParameter()
...
         }
 
+        [Fact]
+        [Trait("spec", "RTC1f")]
+        public async Task WithNullTransportParamsInOptions_ShouldNotThrow()
+        {
+            var client = await GetConnectedClient(options => options.TransportParams = null);
+
+            var uri = LastCreatedTransport.Parameters.GetUri().ToString();
+            Assert.True(true);
```
Comment (sacOO7) [inline src/IO.Ably.Tests.Shared/Realtime/ConnectionSpecs/ConnectionParameterSpecs.cs:148]: 😵😱

File: src/IO.Ably.Shared/Transport/TransportParams.cs [hunk 3]
```diff
@@ -168,6 +175,11 @@ public Uri GetUri()
                 result["clientId"] = ClientId;
             }
 
+            if (AdditionalParameters?.Any() ?? false)
```
Comment (sacOO7) [inline src/IO.Ably.Shared/Transport/TransportParams.cs:178]: Any returns a bool value, can we remove ```?? false``

Comment (sacOO7) [inline src/IO.Ably.Shared/Transport/TransportParams.cs:178]: Can we make it more simple to read
```if ( param is not null && param.any)```
Took me some time to understand why ```false``` was needed

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSpecs/ConnectionParameterSpecs.cs (hunk 2 above)
Comment (marto83) [reply src/IO.Ably.Tests.Shared/Realtime/ConnectionSpecs/ConnectionParameterSpecs.cs:148]: The test just makes sure `GetUri()` doesn't throw an exception.

Comment (sacOO7) [reply src/IO.Ably.Tests.Shared/Realtime/ConnectionSpecs/ConnectionParameterSpecs.cs:148]: Can we use ```RecordException``` to be explicit about assertions 
https://github.com/xunit/xunit/blob/master/test/test.xunit.assert/Asserts/RecordTests.cs

File: src/IO.Ably.Shared/ClientOptions.cs (hunk 1 above)
Comment (marto83) [reply src/IO.Ably.Shared/ClientOptions.cs]: @QuintinWillison I've updated the params to accept `object` for the values. 
@sacOO7 as it's not a very mainstream feature I'm going to keep it simple. :)

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSpecs/ConnectionParameterSpecs.cs (hunk 2 above)
Comment (marto83) [reply src/IO.Ably.Tests.Shared/Realtime/ConnectionSpecs/ConnectionParameterSpecs.cs]: Thanks Sachin. Great suggestion. I updated the code.

File: src/IO.Ably.Shared/Transport/TransportParams.cs (hunk 3 above)
Comment (marto83) [reply src/IO.Ably.Shared/Transport/TransportParams.cs:209]: I haven't upgraded to c# 9 yet. Once we switch we can use `is not null`. Currently it is not available.

Comment (sacOO7) [reply src/IO.Ably.Shared/Transport/TransportParams.cs:209]: what about ``` ( param != null ) ``` 😉

Comment (marto83) [reply src/IO.Ably.Shared/Transport/TransportParams.cs:209]: I think we fall in the lang of personal preference 😁

Comment (sacOO7) [reply src/IO.Ably.Shared/Transport/TransportParams.cs:209]: Yeah, but tbh, this syntax ( sugar ) is new to me : )

Comment (sacOO7) [reply src/IO.Ably.Shared/Transport/TransportParams.cs:209]: See, if you can change it, I am just looking at it from a readability perspective

**Review by sacOO7 — APPROVED** [review]

File: src/IO.Ably.Shared/Transport/TransportParams.cs [hunk 4]
```diff
@@ -106,7 +112,39 @@ internal static async Task<TransportParams> Create(string host, AblyAuth auth, C
             result.UseBinaryProtocol = options.UseBinaryProtocol;
             result.RecoverValue = options.Recover;
             result.Logger = logger ?? options.Logger;
+            result.AdditionalParameters = StringifyParameters(options.TransportParams);
             return result;
+
+            Dictionary<string, string> StringifyParameters(Dictionary<string, object> originalParams)
```
Comment (QuintinWillison) [inline src/IO.Ably.Shared/Transport/TransportParams.cs:118]: I'm not going to insist on it, however my preference would be to lean less on system tests to validate and instead put in place some unit tests specifically for methods like this. It's very unit testable, especially in respect of specific corner cases you had in mind (as is clearly the case looking at the implementation). Perhaps one to think more on for future.

**Review by QuintinWillison — APPROVED** [review]
